Fix geo plot hover interaction for Firefox

LucaVazz opened the pull request:
Related to #5525.

This PR fixes the hover interaction for geo plots (i.e. choropleth) in Firefox and extends the test suite for choropleths to also test for translate3d transforms.

Background:
- The previously used `d3.mouse` depends on the `getScreenCTM` function provided by the browser: https://github.com/plotly/d3/blob/v3.5.18/src/event/mouse.js#L27
- In Firefox, this function is not working as expected compared to other browsers: https://bugzilla.mozilla.org/show_bug.cgi?id=1446011
- D3 does not intend to add a workaround for this problematic behavior: https://github.com/d3/d3/issues/2810
- Therefore we have to use another mean to access the mouse pointer position

LucaVazz commented:
before (v1.58.4): https://codepen.io/lucavazz/pen/mdOpgMY
after fix: https://codepen.io/lucavazz/pen/KKabQpL

Tested with
- Firefox 88.0 on Windows
- Firefox Developer Edition 89.0b3 on Windows
- Chrome 90.0.4430.85 on Windows
- Chrome Dev 92.0.4484.7 on Windows
- Edge  90.0.818.46 on Windows
- Safari 14.0.1 on Mac
- Safari on iPad 14.4.2 
- Firefox 88.1.1 on Android
- Firefox Klar 8.15.1 on Android
- Firefox (Fennec) 87.0.0-rc.1 on Android
- Firefox 88.0.0-beta.6 on Android
- Chrome 90.0.4430.82 on Android
- Chrome Dev 92.0.4484.6 on Android

archmoj commented:
Thanks very much for the information, investigation and PR.
A couple of questions:
 1. Did you investigate why the proposed fix does not work (and appear to break hover) when using Firefox 85 and lower?
 2. `d3.mouse` is used in few other traces and components namely `parcats`, `parcoords` and `sliders`. Could you please investigate if a similar issue existed there? If so, we need to make a centralize function e.g. in `src/lib` to handle the rest.

LucaVazz commented:
Hey, thanks for having a look!

#### 1.
Indeed, I can sadly reproduce it not working for Firefox v <= 85 or the current Firefox ESR (78.10.0esr).
This seems to be due to this bug that was fixed for Firefox 86: https://bugzilla.mozilla.org/show_bug.cgi?id=1684973
The description matches what I observed in testing: Firefox <= 85 sets `offsetX` & `offsetY` relative to the hovered SVG element, while >= 86 sets it relative to the SVG root. Chrome and Safari behave like Firefox >= 86.

What would be the preferred way to proceed? We could test the user agent string and fallback to `d3.mouse` for Firefox <= 85 to have at least non-translated maps working for it.  
Alternatively we could use `layerX` & `layerY`, which are even more non-standard but seem to work fine from a quick test in Firefox ESR. (see https://developer.mozilla.org/en-US/docs/Web/API/UIEvent/layerX )
What's Plotly's general stance on adding code based on the user agent string? Personally I try to avoid it, but it seems like its our only option if we want to get Firefox <= 85 to work with it.

#### 2.
I'll have a look how well the change works with parcats, parcoords and sliders in a bit

archmoj commented:
Given the fact that FireFox >=86 is pretty new, I think we still need to handle versions <=85 as well.
Here are pointers to few locations in plotly.js code that `userAgent` is used to detect certain fallbacks.
https://github.com/plotly/plotly.js/blob/a5a5de98ee43741ae16141ea4f4d32f3d22c9190/src/lib/index.js#L708-L716

https://github.com/plotly/plotly.js/blob/a5a5de98ee43741ae16141ea4f4d32f3d22c9190/src/lib/preserve_drawing_buffer.js#L44-L60

archmoj commented on src/lib/index.js:
```diff
@@ -715,6 +715,18 @@ lib.isIOS = function() {
+    var match = FIREFOX_VERSION_REGEX.exec(window.navigator.userAgent);
+    if(match && match.length === 2) {
+        var versionInt = parseInt(match[1]);
+        if(!Number.isNaN(versionInt)) {
```
Let's use `isNaN` here. Unfortunately `Number.isNaN` is not supported on IE.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isNaN

LucaVazz replied:
Sure, adapted.

archmoj commented on test/jasmine/tests/choropleth_test.js:
```diff
@@ -193,6 +186,18 @@ describe('Test choropleth hover:', function() {
+                assertHoverLabel(0.5 * pos[0], 0.5 * pos[1]);
+            } else {
+                assertHoverLabel(pos[0], pos[1]);
+            }
```
With the latest fix, could we possibly revert changes to this test file now?

LucaVazz replied:
The test changes in general are a nice to have.
However, they are not related to the latest changes. Instead, they increase the coverage by also testing the `translate3d` transform, which might behave differently from the `translate` transform but should still be usable.

Do you have a specific concern against extending the test suite in this regard?

archmoj replied:
Similar tests using `translate` CSS are performed for other trace types.
If you want to increase coverage to take into account `translate3d` that's fine. 
So I think if possible we should keep the previous test as is and add a new one.

LucaVazz replied:
okay, so the goal would be to leave the wrapper as-is so its equal to the wrapper in other tests?

To include testing for `translate3d` than would require to duplicate all 10 test cases that use this wrapper in the following lines if I understand the test setup correctly.

archmoj replied:
Yes. Please let's revert all the changes to this file: `test/jasmine/tests/choropleth_test.js`.

LucaVazz commented:
Added functions for both steps into lib - at first just in the index.js, since I'd argue its not complicated enough to warrant its own file.

Updated the pen and retested including Firefox ESR (78.10.0esr) and works fine for all now.

However, I can't really make sense out of the test failures, when trying it locally it seems all to work.

archmoj commented:
> However, I can't really make sense out of the test failures, when trying it locally it seems all to work.

Some tests are flaky. I re-run the CircleCI and they should pass.

LucaVazz commented:
> Some tests are flaky. I re-run the CircleCI and they should pass.

Thanks, that seems to have helped with the jasmine tests.

What about the other tests / the build? Is this character "conversion" also due to a flaky step...?
![image](https://user-images.githubusercontent.com/4174525/116907280-aa99f400-ac41-11eb-877c-9e4ca3c41d23.png)

archmoj commented:
> > Some tests are flaky. I re-run the CircleCI and they should pass.
> 
> Thanks, that seems to have helped with the jasmine tests.
> 
> What about the other tests / the build? Is this bad character also due to a flaky step...?
> ![image](https://user-images.githubusercontent.com/4174525/116907280-aa99f400-ac41-11eb-877c-9e4ca3c41d23.png)

The bad character problem is really NOT related to your PR. So please do not worry about that.
In fact, the problem appears to be related to non-ASCII characters in d3-v3 `geo` or a memory issue with `node` or `browserify`.
Once we could merge #5112 and cleanup `geo` code from our `d3` fork it should disappear :crossed_fingers:

LucaVazz commented:
Regarding the other trace types:
- `parcats`: The fixed function enables the hover on the bands to work properly in a current Firefox version. In the ESR it fails / shows the hovertext in the top left corner still.  
Example: https://codepen.io/lucavazz/pen/NWdQeaV
- `parcoords`: For this trace I'm unable to produce any hover interaction. Could you please provide an example for it?

I'll have a closer look at the parcats problem in Firefox ESR and the sliders in the coming days.  
Would you prefer to land it all together in one PR?

archmoj commented:
> Regarding the other trace types:
> 
> * `parcats`: The fixed function enables the hover on the bands to work properly in a current Firefox version. In the ESR it fails / shows the hovertext in the top left corner still.
>   unfixed example: https://codepen.io/lucavazz/pen/NWdQeaV
> * `parcoords`: For this trace I'm unable to produce any hover interaction. Could you please provide an example for it?
> 
> I'll have a closer look at the parcats problem in Firefox ESR and the sliders in the coming days.
> Would you prefer to land it all together in one PR?

`parcoords` hover is incomplete. But I noticed a drag issue when trying to move the pink selector on the bottom plots on [this codepen].
If one makes a `d3Mouse` using your patch and use it instead of `d3.mouse` calls, I think that should fix that issue as well. Right?

archmoj commented:
On another note. It appears there are similar bugs related to `geo` handling zoom here:
https://github.com/plotly/plotly.js/blob/a5a5de98ee43741ae16141ea4f4d32f3d22c9190/src/plots/geo/zoom.js#L135
(and if you search for `d3.mouse`) which again could possibly be addressed using your fix around `d3.mouse`.

LucaVazz commented:
- **parcats**: Works reliably in current Firefox (and all other browsers), unfortunately not in Firefox ESR. After a day of investigating I didn't find any indication what is causing this further problem in Firefox ESR. 
Example before fix: https://codepen.io/lucavazz/pen/NWdQeaV
Example after fix: https://codepen.io/lucavazz/pen/qBrWwZN
- **geo zoom**: Unfortunately replacing all d3.mouse calls in the zoom handlers of geo does not fix the problem in Firefox.
As far as I can tell the code paths using the mouse position behave correctly when replacing `d3.mouse` with `Lib.getPositionFromD3Event`. However I don't understand the zoom logic enough to debug the problem.
- **parcoords** and **sliders**: I wan't able to reliably reproduce a problem in Firefox with them. Furthermore, they are out of scope of the original problem.

Unfortunately I'm already over-stretching the time I've allocated for this issue. So I'd appreciate to focus this PR on the geo hover, take on the parcats hover as best effort and record the problems with parcoords and/or sliders into new tickets to be tackled in the future.

archmoj commented:
@LucaVazz thanks very much for the hard work and great contribution.
There is now a conflict in `src/lib/index.js` with `upstream/master`.
Could you please resolve it?

LucaVazz commented:
Thank you as well @archmoj for your great help :)

I've merge master into the fix branch, hopefully its good to go now.

LucaVazz commented:
@archmoj / @alexcjohnson Would this change by the way be a candidate for 2.0.0 or would it likely land in a following release?

archmoj commented:
> @archmoj / @alexcjohnson Would this change by the way be a candidate for 2.0.0 or would it likely land in a following release?

That's a bug we could fix either in 2.0.0 or 2.0.1.
After https://github.com/plotly/plotly.js/pull/5607/commits/ff60f21f881e88395a603d0971efd8b213b198a0 commit a number of tests start failing.  
@LucaVazz would you please fix the problem?
Also it would be nice if you could merge upstream/master one more time into this branch to have more robust tests running on the CI.
Thanks!

archmoj commented:
The failing tests are in `parcats` and `legend`.
```
npm run test-jasmine parcats legend 
```
and using FireFox
```
npm run test-jasmine parcats legend --FF
```
both should pass.

LucaVazz commented:
@archmoj I've merged master and reverted the parcats change. Unfortunately I wasn't able to resolve the failures for parcats, so I'd vote for keeping this PR focused on the fix for geo and to leave parcats in Firefox as-is for now - there seems to be another constraint that I'm unable to fix atm.

archmoj commented:
> @archmoj I've merged master and reverted the parcats change.
> Unfortunately I wasn't able to resolve the failures for parcats, so I'd vote for keeping this PR focused on the fix for geo and to leave parcats in Firefox as-is for now - there seems to be another constraint that I'm unable to fix atm.

@LucaVazz, OK thanks very much. 
I'll have a look :)

archmoj commented on src/lib/index.js:
```diff
@@ -715,6 +715,18 @@ lib.isIOS = function() {
 
+var FIREFOX_VERSION_REGEX = /Firefox\/(\d+)\.\d+/;
+lib.getFirefoxVersion = function() {
+    var match = FIREFOX_VERSION_REGEX.exec(window.navigator.userAgent);
```
It appears regex.`exec` was not used in any other places of our code.
Wondering is it is possible to convert this regex.`exec` to a regex.`test`?

LucaVazz replied:
Unfortunately not, as regex.test does only return true or false, but we need to evaluate the exact version of Firefox in the user agent string.
However, exec has the same compatibility as test: [see mdn](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec#browser_compatibility)

archmoj commented on src/lib/index.js:
```diff
@@ -1267,3 +1279,28 @@ lib.join2 = function(arr, mainSeparator, lastSeparator) {
+ *   at the event relative to the targeted node.
+ */
+lib.getPositionFromD3Event = function() {
+    var firefoxVersion = lib.getFirefoxVersion();
```
We don't want to call `getFirefoxVersion` every time this function is called.
So please declare `firefoxVersion` above `lib.getPositionFromD3Event`.

archmoj replied:
Let's declare `isProblematicFirefox` outside the function.

LucaVazz replied:
good idea, done

archmoj commented on src/lib/index.js:
```diff
@@ -1267,3 +1279,28 @@ lib.join2 = function(arr, mainSeparator, lastSeparator) {
+    var firefoxVersion = lib.getFirefoxVersion();
+
+    // see https://bugzilla.mozilla.org/show_bug.cgi?id=1684973
+    var isProblematicFirefox = firefoxVersion && firefoxVersion < 86;
```
What happens in the case of `firefoxVersion == 0` ?
Perhaps we should say
```js
var isProblematicFirefox = firefoxVersion !== null && firefoxVersion < 86;
```

LucaVazz replied:
fair point, done

archmoj commented:
LGTM.
@alexcjohnson could you please provide a review for this PR as well?

LucaVazz commented on src/lib/index.js:
```diff
@@ -1267,3 +1279,27 @@ lib.join2 = function(arr, mainSeparator, lastSeparator) {
+ * @returns An array with two numbers, representing the x and y coordinates of the mouse pointer
+ *   at the event relative to the targeted node.
+ */
+lib.getPositionFromD3Event = function() {
```
Personally I find ternary operators way harder to read - but I'd say that's just personal preference in the end :D

alexcjohnson replied:
I think the point of @archmoj's comment was not so much about whether to use `if` or a ternary operator, but pulling the conditional out of the function and into the initial script evaluation, for better performance. I agree that ternaries can be confusing when used on such large expressions, but we can do something like:
```js
if (isProblematicFirefox) {
    lib.getPositionFromD3Event = function() {...}
} else {
    lib.getPositionFromD3Event = function() {...}
}
```

alexcjohnson reviewed (approved):
💃 LGTM, with or without the one comment about inverting the function/conditional.

LucaVazz commented:
Thanks for the review @alexcjohnson and thanks again for all the help in getting it integrated properly @archmoj

archmoj commented:
Thanks @LucaVazz for the contribution.
Could you please provide an update on #5525 or if we could close it?
